pkg/client/events/replay_client.go

```diff
@@ -0,0 +1,244 @@
+// observable returns an asynchronous error.
+// This function is intended to be called in a goroutine.
+func (rClient *replayClient[T, R]) goPublishEvents(ctx context.Context) {
+	// React to errors by getting a new events bytes observable, re-mapping it,
+	// and send it to latestObsvblsReplayPublishCh such that
+	// latestObsvbls.Last(ctx, 1) will return it.
+	publishErr := retry.OnError(
+		ctx,
+		eventsBytesRetryLimit,
+		eventsBytesRetryDelay,
+		eventsBytesRetryResetTimeout,
+		"goPublishEvents",
+		rClient.retryPublishEventsFactory(ctx),
+	)
+
+	// If we get here, the retry limit was reached and the retry loop exited.
+	// Since this function runs in a goroutine, we can't return the error to the
+	// caller. Instead, we panic.
+	if publishErr != nil {
+		panic(fmt.Errorf("EventsReplayClient[%T].goPublishEvents should never reach this spot: %w", *new(T), publishErr))
+	}
+}
+
+// retryPublishEventsFactory returns a function which is intended to be passed
+// to retry.OnError. The returned function pipes event bytes from the events
+// query client, maps them to block events, and publishes them to the
+// latestObsvbls replay observable.
+func (rClient *replayClient[T, R]) retryPublishEventsFactory(ctx context.Context) func() chan error {
+	return func() chan error {
+		errCh := make(chan error, 1)
+		eventsBzObsvbl, err := rClient.eventsClient.EventsBytes(ctx, rClient.queryString)
+		if err != nil {
+			errCh <- err
+			return errCh
+		}
+
+		// NB: must cast back to generic observable type to use with Map.
+		// client.BlocksObservable cannot be an alias due to gomock's lack of
+		// support for generic types.
+		eventsBz := observable.Observable[either.Either[[]byte]](eventsBzObsvbl)
+		typedObsrvbl := channel.MapReplay(
+			ctx,
+			latestReplayBufferSize,
+			eventsBz,
+			rClient.newEventsBytesToTypeMapFn(errCh),
+		)
+
+		// Initially set latestObsvbls and update if after retrying on error.
+		rClient.latestObsvblsReplayPublishCh <- typedObsrvbl.(R)
+
+		return errCh
+	}
+}
+
+// newEventsBytesToTypeMapFn is a factory for a function which is intended
+// to be used as a transformFn in a channel.Map() call. Since the map function
+// is called asynchronously, this factory creates a closure around an error
+// channel which can be used for asynchronous error signaling from within the
+// map function, and handling from the Map call context.
+//
+// The map function itself attempts to deserialize the given byte slice as a
+// the EventsReplayClient's generic typed event, using the decoder function provided.
+// If the events bytes observable contained an error, this value is not emitted
+// (skipped) on the destination observable of the map operation.
+//
+// If deserialisation failed because the event bytes were for a different event
+// type, this value is also skipped. If deserialisation failed for some other
+// reason, this function panics.
+func (rClient *replayClient[T, R]) newEventsBytesToTypeMapFn(errCh chan<- error) func(
```

Comment:
Thus far, I've been naming functions that are intended to be used with `channel.Map()` as `mapXXX`. Wdyt about something like this?:

```suggestion
// newMapEventsBytesToTFn is a factory for a function which is intended
// to be used as a transformFn in a channel.Map() call. Since the map function
// is called asynchronously, this factory creates a closure around an error
// channel which can be used for asynchronous error signaling from within the
// map function, and handling from the Map call context.
//
// The map function itself attempts to deserialize the given byte slice as a
// the EventsReplayClient's generic typed event, using the decoder function provided.
// If the events bytes observable contained an error, this value is not emitted
// (skipped) on the destination observable of the map operation.
//
// If deserialisation failed because the event bytes were for a different event
// type, this value is also skipped. If deserialisation failed for some other
// reason, this function panics.
func (rClient *replayClient[T, R]) newMapEventsBytesToTFn(errCh chan<- error) func(
```